File: vercel-ai/template_nextjs/restate/services/agent.ts
import * as restate from "@restatedev/restate-sdk";
import { durableCalls, toolErrorAsTerminalError } from "@restatedev/vercel-ai-middleware"
import { openai } from "@ai-sdk/openai";
import { generateText, tool, wrapLanguageModel } from "ai";
import { z } from "zod";
import { fetchWeather } from "./utils/weather";


// --------------------------------------------------------
//  A simple agent, following the template of the AI SDK
//  tool calling examples
// --------------------------------------------------------

async function simpleAgent(restate: restate.Context, prompt: string) {

  // we wrap the model with the 'durableCalls' middleware, which
  // stores each response in Restate's journal, to be restored on retries
  const model = wrapLanguageModel({
    model: openai("gpt-4o-2024-08-06", { structuredOutputs: true }),
    middleware: durableCalls(restate, { maxRetryAttempts: 3 }),
  });

  const result = await generateText({
    model,
    tools: {
      getWeather: tool({
        description: "Get the current weather for a given city.",
        parameters: z.object({ city: z.string() }),
        execute: async ({ city }) => {
          // call tool wrapped as Restate durable step
          return await restate.run("get weather", () => fetchWeather(city));
        }
      })
    },
    maxSteps: 5,
    // these are local retries by the AI SDK
    // Restate will retry the invocation once those local retries are exhausted to
    // handle longer downtimes, faulty processes, or network communication issues
    maxRetries: 3,
    system: "You are a helpful agent.",
    messages: [{ role: "user", content: prompt }]
  });

  return result.text;

}

// create a simple Restate service as the callable entrypoint
// for our durable agent function
export const agent = restate.service({
  name: "agent",
  handlers: {
    run: async (ctx: restate.Context, prompt: string) => {
      return simpleAgent(ctx, prompt);
    },
  },
  options: {
    journalRetention: { days: 1 },
    ...toolErrorAsTerminalError,
  },
});

export type Agent = typeof agent;

File: integrations/langfuse/agent.py
import restate

from agents import Agent
from pydantic import BaseModel

from utils.context_utils import restate_overrides
from utils.middleware import Runner, function_tool
from utils.utils import fetch_weather, parse_weather_data, WeatherResponse


class WeatherRequest(BaseModel):
    """Request to get the weather for a city."""
    city: str


@function_tool
async def get_weather(req: WeatherRequest) -> WeatherResponse:
    """Get the current weather for a given city."""
    # Do durable steps using the Restate context
    resp = await fetch_weather(req.city)
    return await parse_weather_data(resp)


my_agent = Agent[restate.Context](
    name="Helpful Agent",
    model="gpt-4o",
    instructions="You are a helpful agent.",
    tools=[get_weather],
)


# Agent keyed by conversation id
agent = restate.Service("Agent")

@agent.handler()
async def run(restate_context: restate.Context, message: str) -> str:
    # Your LLM operations as spans under this trace
    with restate_overrides(ctx=restate_context):
        result = await Runner.run(my_agent, input=message)
        return result.final_output

